drivers/sqlite/sqlite.go

```diff
@@ -0,0 +1,321 @@
+	}
+
+	if _, err := os.Stat(sanitizedConnURL); errors.Is(err, os.ErrNotExist) {
+		return nil, &drivers.AppError{Driver: driverName, OrigErr: err, Message: "failed to open db file"}
+	}
+
+	db, err := sql.Open(driverName, sanitizedConnURL)
+	if err != nil {
+		return nil, &drivers.DatabaseError{Driver: driverName, Command: "opening_connection", OrigErr: err, Message: "failed to open connection with the database"}
+	}
+
+	conn, err := db.Conn(context.Background())
+	if err != nil {
+		return nil, &drivers.DatabaseError{Driver: driverName, Command: "grabbing_connection", OrigErr: err, Message: "failed to grab connection to the database"}
+	}
+
+	driverConfig.closeDBonClose = true
+
+	return &sqlite{
+		conn:   conn,
+		db:     db,
+		config: driverConfig,
+	}, nil
+}
+
+func (driver *sqlite) Ping() error {
+	ctx, cancel := drivers.GetContext(driver.config.StatementTimeoutInSecs)
+	defer cancel()
+
+	return driver.conn.PingContext(ctx)
+}
+
+func (sqlite) DriverName() string {
+	return driverName
+}
+
+func (driver *sqlite) Close() error {
+	if driver.conn != nil {
+		if err := driver.conn.Close(); err != nil {
+			return &drivers.DatabaseError{
+				OrigErr: err,
+				Driver:  driverName,
+				Message: "failed to close database connection",
+				Command: "sqlite_conn_close",
+				Query:   nil,
+			}
+		}
+	}
+
+	if driver.db != nil && driver.config.closeDBonClose {
+		if err := driver.db.Close(); err != nil {
+			return &drivers.DatabaseError{
+				OrigErr: err,
+				Driver:  driverName,
+				Message: "failed to close database",
+				Command: "sqlite_db_close",
+				Query:   nil,
+			}
+		}
+		driver.db = nil
+	}
+
+	driver.conn = nil
+	return nil
+}
+
+func (driver *sqlite) Lock() error {
+	return nil
+}
+
+func (driver *sqlite) Unlock() error {
+	return nil
+}
+
+func (driver *sqlite) createSchemaTableIfNotExists() (err error) {
+	ctx, cancel := drivers.GetContext(driver.config.StatementTimeoutInSecs)
+	defer cancel()
+
+	createTableIfNotExistsQuery := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (Version bigint not null primary key, Name varchar not null)", driver.config.MigrationsTable)
+	if _, err = driver.conn.ExecContext(ctx, createTableIfNotExistsQuery); err != nil {
+		return &drivers.DatabaseError{
+			OrigErr: err,
+			Driver:  driverName,
+			Message: "failed while executing query",
+			Command: "create_migrations_table_if_not_exists",
+			Query:   []byte(createTableIfNotExistsQuery),
+		}
+	}
+
+	return nil
+}
+
+func (driver *sqlite) Apply(migration *models.Migration, saveVersion bool) (err error) {
+	query, readErr := migration.Query()
+	if readErr != nil {
+		return &drivers.AppError{
+			OrigErr: readErr,
+			Driver:  driverName,
+			Message: fmt.Sprintf("failed to read migration query: %s", migration.Name),
+		}
+	}
+	defer migration.Close()
+	ctx, cancel := drivers.GetContext(driver.config.StatementTimeoutInSecs)
+	defer cancel()
+
+	if _, err := driver.conn.ExecContext(ctx, query); err != nil {
```

Comment:
We should run this query and the `addMigrationQuery` inside the same transaction